# 2. SVC

In [2]:
from sklearn.svm import SVC
import numpy as np
from keras.datasets import mnist
from tensorflow import keras
from keras import models 
from keras import layers
import numpy as np
import matplotlib.pyplot as plt
from keras.utils import to_categorical



In [3]:
import matplotlib.pyplot as plt
from sklearn import datasets
from sklearn import svm
from sklearn.datasets import fetch_openml

In [4]:
mnist = fetch_openml('mnist_784', version=1)

X, y = mnist['data'],mnist['target']

In [5]:
X.shape, y.shape

((70000, 784), (70000,))

In [6]:
X_test, y_test = X[60000:], y[60000:] # dataset에서 10000개의 testset을 분류

In [7]:
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
%matplotlib inline

X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.3, random_state=42)

In [8]:
import numpy as np
from sklearn.svm import LinearSVC

# Linear SVC will automatically choose one vs all stratagy
lin_clf = LinearSVC(random_state=42)
lin_clf.fit(X_test, y_test)

from sklearn.metrics import accuracy_score

y_pred = lin_clf.predict(X_test)
accuracy_score(y_test, y_pred) #아무런 데이터 전처리 없이 실행했을 때 정확도 약 89%

C:\anaconda3\lib\site-packages\sklearn\svm\_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


0.8945714285714286

아무런 데이터 전처리 없이 실행했을 때 약 89%의 정확도를 보인다.

SVM은 결정경계를 정하는 과정이 있기 때문에 Feature scale을 적절히 맞추어야 한다.(28 X 28 = 784)

In [12]:
from sklearn.preprocessing import StandardScaler # 스케일링 라이브러리
from sklearn.svm import SVC

scaler = StandardScaler()
X_test_scaled = scaler.fit_transform(X_test.astype(np.float32))

In [10]:
lin_clf = LinearSVC(random_state=42)
lin_clf.fit(X_test_scaled,y_test)
y_pred = lin_clf.predict(X_test_scaled)
accuracy_score(y_test,y_pred)
# 데이터 스케일링한 후 정확도 약 93.9%이다.

C:\anaconda3\lib\site-packages\sklearn\svm\_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


0.9398571428571428

# 3. Random Forest

In [13]:
from sklearn.ensemble import RandomForestClassifier

clf = RandomForestClassifier()
clf.fit(X_test,y_test)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [14]:
scaler = StandardScaler()
X_test_scale = scaler.fit_transform(X_test)

clf.fit(X_test_scale, y_test)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [20]:
from sklearn.model_selection import cross_val_score # 교차검증

cross_val_score(clf, X_test, y_test, cv=3, scoring='accuracy')

array([0.95357143, 0.94814286, 0.95557143])

In [21]:
(0.953+0.948+0.955)/3

0.952

Random Forest method로 분류했을 때 : 정확도 약 95.2%
SVC로 분류했을 때 : 정확도 약 94%
Softmax로 분류했을 때 : 정확도 약 99%

input과 output이 동시에 여러개일 때, Softmax로 분류하는 방법이 더 정확함을 확인하였다.